src/main/java/mops/businesslogic/DirectoryServiceImpl.java

```diff
@@ -0,0 +1,253 @@
+    @Override
+    public void checkReadPermission(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+        roleService.checkReadPermission(account, directory);
+    }
+
+    /**
+     * Checks if user has deleting permission.
+     *
+     * @param account user credentials
+     * @param dirId   the id of the folder
+     */
+    @Override
+    public void checkDeletePermission(Account account, long dirId) throws MopsException {
+        Directory directory = fetchDirectory(dirId);
+        roleService.checkDeletePermission(account, directory);
+    }
+
+    /**
+     * Returns all folders of the parent folder.
+     *
+     * @param account     user credentials
+     * @param parentDirID id of the parent folder
+     * @return list of folders
+     */
+    @Override
+    public List<Directory> getSubFolders(Account account, long parentDirID) throws MopsException {
+        Directory directory = fetchDirectory(parentDirID);
+        roleService.checkReadPermission(account, directory);
+        return directoryRepository.getAllSubFoldersOfParent(parentDirID);
+    }
+
+
+    /**
+     * Creates the group root directory.
+     *
+     * @param account user credentials
+     * @param groupId the group id
+     * @return the directory created
+     */
+    @Override
+    public Directory createRootFolder(Account account, Long groupId) throws MopsException {
+
+        roleService.checkIfRole(account, groupId, ADMINISTRATOR);
+        Set<String> roleNames = permissionService.fetchRolesInGroup(groupId);
+        Set<DirectoryPermissionEntry> permissions = createDefaultPermissions(roleNames);
+        DirectoryPermissions permission = new DirectoryPermissions(permissions);
+        DirectoryPermissions rootPermissions = directoryPermissionsRepo.save(permission);
+        Directory directory = Directory.of(groupId.toString(), null, groupId, rootPermissions);
+        return directoryRepository.save(directory);
+    }
+
+    /**
+     * Creates a new folder inside a folder.
+     *
+     * @param account     user credentials
+     * @param parentDirId id of the parent folder
+     * @param dirName     name of the new folder
+     * @return id of the new folder
+     */
+    @Override
+    public Directory createFolder(Account account, long parentDirId, String dirName) throws MopsException {
+        Directory rootDirectory = fetchDirectory(parentDirId);
+        roleService.checkWritePermission(account, rootDirectory);
+        Directory directory = rootDirectory.createSubDirectory(dirName); //NOPMD// this is no violation of demeter's law
+        return directoryRepository.save(directory);
+    }
+
+    /**
+     * Deletes a folder.
+     *
+     * @param account user credential
+     * @param dirId   id of the folder to be deleted
+     * @return the parent id of the deleted folder
+     */
+    @Override
+    @SuppressWarnings("PMD.LawOfDemeter") //these are not violations of demeter's law
```

Comment:
good, you did it here, why not everywhere else?

Reply:
yeah yeah